src/components/sections/Filter.js

```diff
@@ -30,6 +32,27 @@ const Filter = () => {
     [dispatch],
   );
 
+  // get edge creation and display values for edges across all prompts
+  let shouldShowWarning = false;
+  const prompts = useSelector((state) => getFormValue(state, 'prompts'));
+
+  if (prompts) {
+    const edgeCreationValues = prompts
+      .filter((prompt) => prompt?.edges?.create)
+      .map((prompt) => prompt.edges.create);
+
+    const edgeDisplayValues = prompts
+      .filter((prompt) => prompt?.edges?.display)
+      .flatMap((prompt) => prompt.edges.display);
+
+    if (edgeCreationValues.length > 0 || edgeDisplayValues.length > 0) {
+      shouldShowWarning = getEdgeFilteringWarning(
+        currentValue.rules,
+        [...edgeCreationValues, ...edgeDisplayValues],
+      );
+    }
+  }
+
```

Comment:
While this works, there are a few rough edges with it:
- The `edgeCreationValues` and `edgeDisplayValues` variables (derived state) aren't memoized, so they will be recomputed on every render. This isn't terrible in this situation because the computation is relatively cheap, but we can explicitly tell React when the values need to be recomputed using `useMemo`. Note that this will possibly be automatic once we can use the React compiled in React 19.
- Similarly, the way you have set up `shouldShowWarning` isn't a particularly "React-y" way of doing things. Try to think of it in terms of "a computed value that must update under certain circumstances", rather than the way you have it now which is more of a vanilla JS way of thinking about it as "a variable that I will update the value of". Not quite sure I am explaining this well, but what this comes back to is the often quoted Dan Abramov line `ui = fn(data)(state)`. 

So you could address those issues like this:

```suggestion
  const prompts = useSelector((state) => getFormValue(state, "prompts"));
  
  const { edgeCreationValues, edgeDisplayValues } = useMemo(() => {
    if (!prompts) return { edgeCreationValues: [], edgeDisplayValues: [] };
    const edgeCreationValues = [];
    const edgeDisplayValues = [];

    prompts.forEach((prompt) => {
      if (prompt?.edges?.create) edgeCreationValues.push(prompt.edges.create);
      if (prompt?.edges?.display) edgeDisplayValues.push(...prompt.edges.display);
    });

    return { edgeCreationValues, edgeDisplayValues };
  }, [prompts]);

  const shouldShowWarning = useMemo(() => {
    if (edgeCreationValues.length > 0 || edgeDisplayValues.length > 0) {
      return getEdgeFilteringWarning(currentValue.rules, [...edgeCreationValues, ...edgeDisplayValues]);
    }
    return false;
  }, [currentValue, edgeCreationValues, edgeDisplayValues]);
```

An alternative would be to make a selector that encapsulates all of this:

```
const getFormValue = formValueSelector('edit-stage');
const getPrompts = (state) => getFormValue(state, 'prompts');
const getCurrentValue = (state) => getFormValue(state, 'filter');

export const getShouldShowEdgeWarning = createSelector(
  [getPrompts, getCurrentValue],
  (prompts, currentValue) => {
    if (!prompts) return { edgeCreationValues: [], edgeDisplayValues: [], shouldShowWarning: false };

    const edgeCreationValues = [];
    const edgeDisplayValues = [];

    prompts.forEach((prompt) => {
      if (prompt?.edges?.create) edgeCreationValues.push(prompt.edges.create);
      if (prompt?.edges?.display) edgeDisplayValues.push(...prompt.edges.display);
    });

    const shouldShowWarning = (edgeCreationValues.length > 0 || edgeDisplayValues.length > 0) 
      ? getEdgeFilteringWarning(currentValue.rules, [...edgeCreationValues, ...edgeDisplayValues])
      : false;

    return shouldShowWarning;
  }
);
```

I think conceptually speaking, it is less readable to derive values inside a selector like this. If you wanted to encapsulate the logic in a single place, a custom hook would probably be more appropriate. However, since this logic isn't reused, it doesn't really make sense to go to that trouble.

Reply:
resolved with first approach [3422a77](https://github.com/complexdatacollective/Architect/pull/840/commits/3422a77b6b5bc3fa4cfd493966982cf82585eafa)